src/callbacks.cpp

```diff
@@ -0,0 +1,231 @@
+ * updates the mission progress of the WaypointNode based on the received
+ * message.
+ *
+ * @param msg The mission progress message received.
+ */
+void WaypointNode::callback_mission_progress(
+    const interfaces::msg::MissionProgress &msg) {
+    // Store current timestamp for later
+    rclcpp::Time timestamp_now = this->now();
+
+    // Ignore message if node is not active
+    if (!this->get_active()) return;
+
+    RCLCPP_DEBUG(this->get_logger(),
+                 "WaypointNode::%s: Received mission "
+                 "progress from '%s': progress: %f / 1.0",
+                 __func__, msg.sender_id.c_str(), msg.progress);
+
+    // Check timestamp
+    if (timestamp_now - rclcpp::Time(msg.time_stamp) >
+        rclcpp::Duration(std::chrono::duration<int64_t, std::milli>(
+            max_progress_msg_time_difference_ms))) {
+        // Warn
+        RCLCPP_WARN(
+            get_logger(),
+            "WaypointNode::%s: Received too "
+            "old timestamp in progress message: %s. Using value regardless.",
+            __func__, msg.sender_id.c_str());
+    }
+
+    // Store value
+    mission_progress = msg.progress;
+}
+
+/**
+ * @brief Callback function for handling wait time completion.
+ *
+ * This function is called when the wait time for a waypoint is completed.
+ * It deactivates the timer, checks if the mission was aborted during the wait
+ * time, and reacts according to the current state of the node.
+ *
+ * @note If the node has an incorrect state, a fatal error is logged and a
+ * job_finished error message is sent.
+ */
+void WaypointNode::callback_wait_time() {
+    // Deactivate Timer
+    wait_timer->cancel();
+
+    // Check if the mission was aborted during the wait time
+    if (!this->get_active()) return;
+
+    // React according to state
+    switch (get_node_state()) {
+        case pre_wait_time:
+            set_node_state(reach_cruise_height);
+            RCLCPP_INFO(this->get_logger(),
+                        "WaypointNode::%s: '%s' finished. New state: '%s'",
+                        __func__, get_node_state_str(pre_wait_time),
+                        get_node_state_str());
+            break;
+        case post_wait_time:
+            RCLCPP_INFO(this->get_logger(),
+                        "WaypointNode::%s: '%s' finished. Deactivating node.",
+                        __func__, get_node_state_str());
+
+            // Job finished successfully: Reset node for next command
+            RCLCPP_INFO(this->get_logger(),
+                        "WaypointNode::%s: --- NODE DEACTIVATED ---", __func__);
+
+            this->job_finished();
+            reset_node();
+            break;
+        default:
+            RCLCPP_FATAL(this->get_logger(),
+                         "WaypointNode::%s: Node has incorrect "
+                         "state: %d",
+                         __func__, get_node_state());
+            this->job_finished(
+                "WaypointNode::" + (std::string) __func__ +
```

Comment:
Would start at the beginning:
```suggestion
                std::string("WaypointNode::") + __func__ +
```

Reply:
Agree